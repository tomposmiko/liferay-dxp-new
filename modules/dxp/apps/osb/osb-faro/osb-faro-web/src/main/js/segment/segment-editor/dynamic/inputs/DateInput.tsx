import autobind from 'autobind-decorator';
import DateInput from 'shared/components/DateInput';
import Form from 'shared/components/form';
import moment from 'moment';
import React from 'react';
import {INPUT_DATE_FORMAT, PropertyTypes} from '../utils/constants';
import {ISegmentEditorInputBase} from '../utils/types';

interface IEditorDateInputProps extends ISegmentEditorInputBase {
	displayFormat: string;
	value: string;
}

export default class EditorDateInput extends React.Component<IEditorDateInputProps> {
	componentDidMount() {
		const {
			id,
			property: {entityName, type},
			valid
		} = this.props;

		if (!id && valid) {
			analytics.track('Dynamic Segment Creation - Completed Attribute', {
				entityName,
				type
			});
		}
	}

	@autobind
	handleDateChange(value) {
		this.props.onChange({
			type: PropertyTypes.Date,
			value
		});
	}

	render() {
		const {
			className,
			displayFormat,
			displayValue,
			operatorRenderer: OperatorDropdown,
			property: {entityName},
			value
		} = this.props;

		const date = moment(value, INPUT_DATE_FORMAT).format(INPUT_DATE_FORMAT);

		return (
			<div className='criteria-statement'>
				<Form.Group autoFit>
					<Form.GroupItem className='entity-name' label shrink>
						{entityName}
					</Form.GroupItem>

					<Form.GroupItem className='display-value' label shrink>
						{displayValue}
					</Form.GroupItem>

					<OperatorDropdown />

					<Form.GroupItem shrink>
						<DateInput
							className={className}
							displayFormat={displayFormat}
							onDateInputChange={this.handleDateChange}
							value={date}
						/>
					</Form.GroupItem>
				</Form.Group>
			</div>
		);
	}
}
