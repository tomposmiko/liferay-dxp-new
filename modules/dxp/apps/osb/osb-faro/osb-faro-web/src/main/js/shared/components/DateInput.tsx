import Button from './Button';
import Card from './Card';
import DatePicker from './date-picker';
import getCN from 'classnames';
import Icon from './Icon';
import Input from './Input';
import MaskedInput from './MaskedInput';
import moment from 'moment';
import Overlay from './Overlay';
import React, {useState} from 'react';
import {
	applyTimeZone,
	DATE_MASK,
	DATE_TIME_MASK,
	FORMAT
} from 'shared/util/date';
import {noop} from 'lodash';

interface IDateInputProps extends React.HTMLAttributes<HTMLInputElement> {
	displayFormat?: string;
	format?: string;
	name?: string;
	onDateInputBlur?: (param: any) => void;
	onDateInputChange?: (param: any) => void;
	overlayAlignment?: OverlayAlignment;
	readOnly?: boolean;
	showTimeSelector?: boolean;
	timeZoneId?: string;
	usePortal?: boolean;
	value: any;
}

export type OverlayAlignment = 'bottomLeft' | 'rightCenter';

const DateInput: React.FC<IDateInputProps> = ({
	className,
	displayFormat,
	format = FORMAT,
	id,
	name,
	onDateInputBlur = noop,
	onDateInputChange = noop,
	overlayAlignment = 'bottomLeft',
	readOnly = false,
	showTimeSelector = false,
	timeZoneId,
	usePortal = true,
	value
}) => {
	const [active, setActive] = useState(false);

	const handleClick = (event: PointerEvent): void => {
		event.preventDefault();

		setActive(!active);
	};

	const handleDateSelect = (value: any): void => {
		onDateInputChange(value.format(format));
	};

	const handleChange = ({target}: React.BaseSyntheticEvent<Event>): void => {
		const {value} = target;

		if (moment(value, format, true).isValid()) {
			onDateInputChange(value);
		}
	};

	const handleOutsideClick = (event: PointerEvent): void => {
		if (onDateInputBlur && active) {
			onDateInputBlur(event);
		}

		setActive(false);
	};

	const getDateValue = (): string => {
		if (!displayFormat) {
			return value;
		}

		let date = value;

		if (showTimeSelector) {
			date = applyTimeZone(value, timeZoneId);
		} else {
			date = moment(value);
		}

		return date.format(displayFormat);
	};

	const date = showTimeSelector
		? applyTimeZone(value, timeZoneId)
		: moment(value, format);

	return (
		<Overlay
			active={active}
			alignment={overlayAlignment}
			className={getCN('date-input-root', className)}
			forceAlignment={false}
			onOutsideClick={handleOutsideClick}
			usePortal={usePortal}
		>
			<Input.Group>
				<Input.GroupItem>
					<MaskedInput
						autoComplete='off'
						data-testid='date-input'
						id={id}
						inset='after'
						keepCharPositions
						mask={showTimeSelector ? DATE_TIME_MASK : DATE_MASK}
						name={name}
						onChange={handleChange}
						onClick={handleClick}
						placeholder={
							showTimeSelector
								? Liferay.Language.get('yyyy-mm-dd-hh-mm-zz')
								: Liferay.Language.get('yyyy-mm-dd')
						}
						readOnly={readOnly}
						showMask
						value={getDateValue()}
					/>

					<Input.Inset position='after'>
						<Button display='unstyled' onClick={handleClick}>
							<Icon symbol='calendar' />
						</Button>
					</Input.Inset>
				</Input.GroupItem>
			</Input.Group>

			<Card>
				<Card.Body>
					<DatePicker
						date={date.isValid() ? date : null}
						minDate={moment().subtract(100, 'years')}
						onSelect={handleDateSelect}
						showTimeSelector={showTimeSelector}
						timeZoneId={timeZoneId}
					/>
				</Card.Body>
			</Card>
		</Overlay>
	);
};

export default DateInput;
