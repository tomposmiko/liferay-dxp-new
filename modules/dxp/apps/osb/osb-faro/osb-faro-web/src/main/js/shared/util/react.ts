import {isEqual} from 'lodash';

export const getDisplayName = WrappedComponent =>
	WrappedComponent.displayName || WrappedComponent.name || 'Component';

type HasChanges = <T>(
	prev: T | object,
	next: T | object,
	...keys: string[]
) => boolean;

/**
 * Compare previous state or props object by provided keys to detect changes.
 */
export const hasChanges: HasChanges = (prev = {}, next = {}, ...keys) => {
	for (const key of keys) {
		if (key in next) {
			const newVal = next[key];

			const prevVal = prev[key];

			if (!isEqual(newVal, prevVal)) {
				return true;
			}
		}
	}

	return false;
};
