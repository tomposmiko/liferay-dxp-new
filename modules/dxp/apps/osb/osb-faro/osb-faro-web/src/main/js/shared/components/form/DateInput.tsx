import DateInput, {OverlayAlignment} from '../DateInput';
import getCN from 'classnames';
import HelpBlock from './HelpBlock';
import Label from './Label';
import React from 'react';
import {FieldProps} from 'formik';
import {isNumber} from 'lodash';

interface IFormDateInputProps
	extends FieldProps,
		React.HTMLAttributes<HTMLElement> {
	inline: boolean;
	label: string;
	overlayAlignment?: OverlayAlignment;
	popover: {
		content: React.ReactNode;
		title: React.ReactNode;
	};
	required?: boolean;
	usePortal?: boolean;
	width: number;
}

const FormDateInput: React.FC<IFormDateInputProps> = ({
	className,
	field,
	form,
	inline = false,
	label,
	overlayAlignment = 'bottomLeft',
	popover,
	required = false,
	usePortal = true,
	width
}) => {
	const {name, value} = field;

	const handleChange = (value): void => {
		const {setFieldValue} = form;

		setFieldValue(name, value);
	};

	const error = form.errors[name];
	const touched = form.touched[name];

	const classes = getCN('form-date-input-root', className, {
		'form-inline-group': inline,
		'has-error': error && touched,
		'has-success': !error && touched
	});

	const style = isNumber(width)
		? {flexBasis: `${width}%`, flexGrow: 0}
		: undefined;

	return (
		<div className={classes} style={style}>
			{label && (
				<Label htmlFor={name} popover={popover} required={required}>
					{label}
				</Label>
			)}

			<DateInput
				id={name}
				name={name}
				onDateInputChange={handleChange}
				overlayAlignment={overlayAlignment}
				usePortal={usePortal}
				value={value}
			/>

			<HelpBlock name={name} />
		</div>
	);
};

export default FormDateInput;
