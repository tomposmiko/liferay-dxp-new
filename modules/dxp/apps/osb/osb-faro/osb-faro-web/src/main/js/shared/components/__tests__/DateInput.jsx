import DateInput from '../DateInput';
import React from 'react';
import {render} from '@testing-library/react';

jest.unmock('react-dom');

describe('DateInput', () => {
	it('should render', () => {
		const {container} = render(<DateInput />);
		expect(container).toMatchSnapshot();
	});

	it('should use the displayFormat prop for displaying the date', () => {
		const {getByDisplayValue} = render(
			<DateInput
				displayFormat='YYYY MM DD HH:mm'
				onDateInputChange={jest.fn()}
				value='1970-01-01'
			/>
		);

		expect(getByDisplayValue('1970 01 01 00:00')).toBeTruthy();
	});
});
