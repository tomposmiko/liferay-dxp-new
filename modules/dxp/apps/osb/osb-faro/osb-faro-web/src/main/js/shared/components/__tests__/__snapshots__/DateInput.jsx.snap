// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateInput should render 1`] = `
<div>
  <div
    aria-haspopup="true"
    class="input-group date-input-root"
  >
    <div
      class="input-group-item"
    >
      <input
        autocomplete="off"
        class="input-root form-control input-group-inset input-group-inset-after"
        data-testid="date-input"
        placeholder="YYYY-MM-DD"
        type="text"
        value=""
      />
      <div
        class="input-group-inset-item input-group-inset-item-after"
      >
        <button
          class="button-root btn btn-unstyled"
          type="button"
        >
          <svg
            class="lexicon-icon lexicon-icon-calendar icon-root"
            role="presentation"
          >
            <use
              xlink:href="#calendar"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;
